File: src/test/runner.ts
import { readFileSync, readdirSync, existsSync, watch, writeFileSync, mkdirSync, rmSync } from "fs";
import { execSync } from "child_process";
import { join, relative } from "path";
import * as os from "os";
import { Lexer } from "../lex/lexer";
import { Parser } from "../parse/parser";
import { Engine } from "../diag/engine";
import { Resolver } from "../sema/resolver";
import { TypeChecker } from "../types/checker";
import { lower } from "../lower/lower";
import { build } from "../backend/zig/driver";
import { getPin, detect } from "../toolchain/manager";
import * as ast from "../ast/node";

export interface TestSuite {
  name: string;
  file: string;
  line: number;
}

export interface TestResult {
  suite: TestSuite;
  passed: boolean;
  timeMs: number;
  output: string;
  error: string;
}

/**
 * Discovers all `.rt` files in the given root and extracts `test "name" {}` blocks.
 */
export function discoverTests(root: string): TestSuite[] {
  const suites: TestSuite[] = [];
  
  function walk(dir: string) {
    if (!existsSync(dir)) return;
    const entries = readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
      const path = join(dir, entry.name);
      if (entry.isDirectory() && entry.name !== "node_modules" && entry.name !== ".tratio-build" && entry.name !== ".tratio-test") {
        walk(path);
      } else if (entry.name.endsWith(".rt")) {
        const src = readFileSync(path, "utf8");
        const diag = new Engine();
        const lexer = new Lexer(src, diag);
        const toks = lexer.lex();
        const parser = new Parser(toks, diag);
        const program = parser.parse();
        
        for (const mod of program.mods) {
          for (const stmt of mod.body.stmts) {
            if (stmt.tag === "test") {
              suites.push({
                name: (stmt as ast.TestBlock).name.value,
                file: path,
                line: stmt.span.line
              });
            }
          }
        }
      }
    }
  }
  
  walk(root);
  return suites;
}

/**
 * Synthesizes a `start()` entry point for modules containing `test` blocks.
 * Converts `test "name" { ... }` into private functions and calls them sequentially.
 * If a `start()` function already exists, it prepends the test calls to it.
 */
function synthesizeTests(program: ast.Program): void {
  for (const mod of program.mods) {
    const testBlocks = mod.body.stmts.filter((s): s is ast.TestBlock => s.tag === "test");
    if (testBlocks.length === 0) continue;

    // 1. Convert test blocks to private functions
    const testFuncs: ast.Function[] = testBlocks.map((tb, i) => ({
      span: tb.span,
      tag: "function",
      public: false,
      name: { span: tb.span, tag: "name", text: `__test_${i}` },
      params: [],
      body: tb.body
    }));

    // 2. Create statements to call these test functions
    const callStmts: ast.Statement[] = testFuncs.map((tf) => ({
      span: tf.span,
      tag: "action",
      name: { span: tf.span, tag: "name", text: tf.name.text },
      args: []
    }));

    // 3. Check if `start` already exists
    const existingStart = mod.body.stmts.find(
      s => s.tag === "function" && (s as ast.Function).name.text === "start"
    ) as ast.Function | undefined;

    if (existingStart) {
      // Prepend test calls to the existing start function's body
      existingStart.body.stmts.unshift(...callStmts);
    } else {
      // Create a new public start function
      const dummySpan = { start: 0, end: 0, line: 1, col: 1 };
      const startFunc: ast.Function = {
        span: dummySpan,
        tag: "function",
        public: true,
        name: { span: dummySpan, tag: "name", text: "start" },
        params: [],
        body: { span: dummySpan, tag: "block", stmts: callStmts }
      };
      mod.body.stmts.push(...testFuncs, startFunc);
    }

    // 4. Remove the original test block statements from the module body
    mod.body.stmts = mod.body.stmts.filter(s => s.tag !== "test");
  }
}

/**
 * Compiles and executes a single test file.
 */
function runSingleTest(suite: TestSuite, projectRoot: string): TestResult {
  const startMs = Date.now();
  const src = readFileSync(suite.file, "utf8");
  const diag = new Engine();
  
  // 1. Parse
  const lexer = new Lexer(src, diag);
  const toks = lexer.lex();
  const parser = new Parser(toks, diag);
  let program = parser.parse();
  
  if (diag.failed) {
    return { suite, passed: false, timeMs: Date.now() - startMs, output: "", error: "Parse failed" };
  }

  // 2. Synthesize test entry points
  synthesizeTests(program);

  // 3. Resolve
  const resolver = new Resolver(diag, projectRoot);
  resolver.resolve(program, "test_runner");
  if (diag.failed) {
    return { suite, passed: false, timeMs: Date.now() - startMs, output: "", error: "Resolution failed" };
  }

  // 4. Type Check
  const checker = new TypeChecker(diag);
  checker.check(program);
  if (diag.failed) {
    return { suite, passed: false, timeMs: Date.now() - startMs, output: "", error: "Type check failed" };
  }

  // 5. Lower to IR
  const irMod = lower(program);

  // 6. Build
  const outDir = join(os.tmpdir(), `tratio-test-${Date.now()}-${Math.random().toString(36).slice(2)}`);
  mkdirSync(outDir, { recursive: true });
  
  const pinnedVersion = getPin() || "0.16.0";
  const toolchain = detect(pinnedVersion) || { version: pinnedVersion, path: "zig" };
  
  try {
    const buildResult = build(irMod, "dev", outDir, toolchain.path);
    if (!buildResult.success) {
      return { suite, passed: false, timeMs: Date.now() - startMs, output: "", error: buildResult.error || "Build failed" };
    }

    // 7. Execute
    const execResult = execSync(join(outDir, "app"), { 
      encoding: "utf8", 
      maxBuffer: 10 * 1024 * 1024,
      stdio: ["pipe", "pipe", "pipe"]
    });
    
    // Cleanup
    rmSync(outDir, { recursive: true, force: true });

    return { 
      suite, 
      passed: true, 
      timeMs: Date.now() - startMs, 
      output: execResult.toString().trim(), 
      error: "" 
    };
  } catch (e: any) {
    // Cleanup on failure
    if (existsSync(outDir)) {
      rmSync(outDir, { recursive: true, force: true });
    }
    return { 
      suite, 
      passed: false, 
      timeMs: Date.now() - startMs, 
      output: e.stdout || "", 
      error: e.stderr || e.message 
    };
  }
}

/**
 * Orchestrates test discovery, execution, reporting, and watch mode.
 */
export function runTests(
  root: string = process.cwd(), 
  filter?: string, 
  watchMode: boolean = false, 
  junitPath?: string
): void {
  let suites = discoverTests(root);
  
  // If no explicit `test` blocks were found, fallback to running any `.rt` file 
  // in the `tests/` directory that has a `start()` function (e.g., stdlib tests).
  if (suites.length === 0 && existsSync(join(root, "tests"))) {
    const testFiles = readdirSync(join(root, "tests"), { recursive: true })
      .filter((f: any) => f.toString().endsWith(".rt"))
      .map((f: any) => join(root, "tests", f.toString()));
    
    suites = testFiles.map((file: string) => ({
      name: file.split("/").pop()!.replace(".rt", ""),
      file,
      line: 1
    }));
  }

  if (filter) {
    suites = suites.filter(s => s.name.includes(filter));
  }
  
  console.log(`Found ${suites.length} test suites.`);
  
  const run = () => {
    let passed = 0;
    let failed = 0;
    const results: any[] = [];
    
    for (const suite of suites) {
      const result = runSingleTest(suite, root);
      
      if (result.passed) {
        console.log(`  ✓ ${result.suite.name} (${relative(root, result.suite.file)}:${result.suite.line}) [${result.timeMs}ms]`);
        passed++;
      } else {
        console.log(`  ✗ ${result.suite.name} (${relative(root, result.suite.file)}:${result.suite.line})`);
        console.log(`    Error: ${result.error}`);
        if (result.output) {
          console.log(`    Output: ${result.output}`);
       0}
        failed++;
      }
      
      results.push({ 
        name: result.suite.name, 
        classname: relative(root, result.suite.file), 
        time: (result.timeMs / 1000).toFixed(3),
        failure: result.passed ? undefined : result.error
      });
    }
    
    console.log(`\n${passed} passed, ${failed} failed.`);
    
    if (junitPath) {
      const xml = `<?xml version="1.0" encoding="UTF-8"?>
<testsuite name="tratio" tests="${passed + failed}" failures="${failed}">
${results.map(r => `  <testcase name="${r.name}" classname="${r.classname}" time="${r.time}">${r.failure ? `<failure message="${r.failure.replace(/"/g, '&quot;')}"/>` : ""}</testcase>`).join("\n")}
</testsuite>`;
      writeFileSync(junitPath, xml);
      console.log(`JUnit report written to ${junitPath}`);
    }

    if (failed > 0) {
      process.exitCode = 1;
    }
  };
  
  run();
  
  if (watchMode) {
    console.log("\nWatching for changes...");
    watch(root, { recursive: true }, (event, filename) => {
      if (filename && filename.endsWith(".rt")) {
        console.log(`\nChange detected: ${filename}`);
        // Reset exit code for watch mode re-runs
        process.exitCode = 0;
        run();
      }
    });
  }
}

File: src/test/runner.nim
import std/[json, os, osproc, sequtils, strutils, tables, times]
import ../ast/node as ast
import ../diag/engine
import ../diag/render
import ../lex/lexer
import ../parse/parser
import ../sema/resolver
import ../sema/link
import ../interop/expand
import "../interop/bind.nim" as cBinding
import ../pkg/lint
import ../types/checker
import ../ir/[lower, monomorph]
import ../backend/zig/driver as zigDriver
import ../backend/c/driver as cDriver
import ../build/options as buildOptions
import ../toolchain/manager

type
  TestSuite* = object
    name*: string
    file*: string
    line*: int
  TestResult* = object
    suite*: TestSuite
    passed*: bool
    timeMs*: int
    output*: string
    error*: string
    files*: seq[string]
    opaque*: bool
  TestExecutor* = proc(suite: TestSuite): TestResult {.closure.}
  TestWatcher* = ref object
    root*: string
    filter*: string
    backend*: string
    executor*: TestExecutor
    stopped*: bool
    snapshot*: Table[string, string]

proc parseFile(path: string): tuple[program: ast.Program, failed: bool] =
  let source = readFile(path)
  let diag = newEngine()
  diag.setSource(source, path)
  let program = newParser(newLexer(source, diag).lex(), diag).parse()
  (program: program, failed: diag.failed)

proc inspect(path: string; includeStart: bool; suites: var seq[TestSuite]) =
  if not path.toLowerAscii().endsWith(".iv"): return
  let parsed = parseFile(path)
  if parsed.failed:
    suites.add(TestSuite(name: splitFile(path).name, file: path, line: 1))
    return
  for unit in parsed.program.units:
    var hasTest = false
    for statement in unit.body.stmts:
      if statement.tag == "test":
        hasTest = true
        let testBlock = ast.TestBlock(statement)
        suites.add(TestSuite(name: testBlock.name.value, file: path, line: statement.span.line))
    if includeStart and not hasTest:
      for statement in unit.body.stmts:
        if statement.tag == "function" and ast.Function(statement).name.text == "start":
          suites.add(TestSuite(name: splitFile(path).name, file: path, line: statement.span.line))
          break

proc discoverTests*(root: string; includeStarts = false): seq[TestSuite] =
  var stdlib = ""
  for candidate in [getAppDir().parentDir / "test" / "stdlib",
      getAppDir().parentDir.parentDir / "test" / "stdlib",
      currentSourcePath.parentDir.parentDir.parentDir / "test" / "stdlib"]:
    if dirExists(candidate):
      stdlib = absolutePath(candidate)
      break
  let selected = if root == "std": stdlib else: absolutePath(root)
  let starts = includeStarts or fileExists(selected) or selected == stdlib
  var suites: seq[TestSuite] = @[]
  if fileExists(selected):
    inspect(selected, true, suites)
    return suites
  if not dirExists(selected): return @[]
  proc visit(directory: string) =
    for kind, path in walkDir(directory):
      let name = path.lastPathPart
      if kind == pcLinkToDir or kind == pcLinkToFile: continue
      if kind == pcDir:
        if name in ["node_modules", ".artifacts", ".git"]: continue
        visit(path)
      elif kind == pcFile:
        inspect(path, starts, suites)
  visit(selected)
  result = suites

proc synthesizeTests*(program: ast.Program; selected: TestSuite) : bool =
  ## Replace the selected test block with a start entry that calls only that block.
  for unit in program.units:
    var target: ast.TestBlock
    for statement in unit.body.stmts:
      if statement.tag == "test":
        let testBlock = ast.TestBlock(statement)
        if testBlock.name.value == selected.name and statement.span.line == selected.line:
          target = testBlock
          break
    if target == nil: continue
    let span = target.span
    let nameNode = ast.Name(tag: "name", span: span, text: "__test_selected")
    let function = ast.Function(tag: "function", span: span, public: false, name: nameNode,
      typeParams: @[], params: @[], returnType: nil, abi: "", constraint: nil,
      constraints: @[], attributes: @[], body: target.body)
    let callName = ast.Name(tag: "name", span: span, text: nameNode.text)
    let call = ast.Action(tag: "action", span: span, name: callName, args: @[], value: nil)
    var start: ast.Function
    for statement in unit.body.stmts:
      if statement.tag == "function" and ast.Function(statement).name.text == "start":
        start = ast.Function(statement)
        break
    if start == nil:
      let startName = ast.Name(tag: "name", span: span, text: "start")
      start = ast.Function(tag: "function", span: span, public: true, name: startName,
        typeParams: @[], params: @[], returnType: nil, abi: "", constraint: nil,
        constraints: @[], attributes: @[], body: ast.Block(tag: "block", span: span, stmts: @[ast.Statement(call)]))
      unit.body.stmts.add(ast.Statement(start))
    else:
      start.body.stmts = @[ast.Statement(call)]
    unit.body.stmts.add(ast.Statement(function))
    var filtered: seq[ast.Statement] = @[]
    for statement in unit.body.stmts:
      if statement != ast.Statement(target): filtered.add(statement)
    unit.body.stmts = filtered
    return true
  for unit in program.units:
    for statement in unit.body.stmts:
      if statement.tag == "function" and
          ast.Function(statement).name.text == "start" and
          statement.span.line == selected.line:
        return true

proc strings(node: JsonNode): seq[string] =
  if node != nil and node.kind == JArray:
    for value in node:
      if value.kind == JString: result.add(value.getStr())

proc child(node: JsonNode; name: string): JsonNode =
  if node != nil and node.kind == JObject and node.hasKey(name): node[name]
  else: nil

proc text(node: JsonNode; fallback = ""): string =
  if node != nil and node.kind == JString: node.getStr() else: fallback

proc flag(node: JsonNode): bool =
  node != nil and node.kind == JBool and node.getBool()

proc rooted(root, path: string): string =
  if isAbsolute(path): path else: absolutePath(root / path)

proc runSingleTest*(suite: TestSuite; backend = "zig"; executor: TestExecutor = nil): TestResult =
  let started = epochTime()
  if executor != nil:
    result = executor(suite)
    result.timeMs = int((epochTime() - started) * 1000)
    return
  result.suite = suite
  try:
    var projectRoot = if dirExists(suite.file): suite.file else: parentDir(suite.file)
    while not fileExists(projectRoot / "project.json") and
        parentDir(projectRoot) != projectRoot:
      projectRoot = parentDir(projectRoot)
    if not fileExists(projectRoot / "project.json"):
      projectRoot = parentDir(suite.file)
    let manifestPath = projectRoot / "project.json"
    let manifest = parseJson(if fileExists(manifestPath): readFile(manifestPath) else: "{}")
    let build = manifest.child("build")
    let c = build.child("c")
    let linkOptions = build.child("link")
    let suiteOptions = manifest.child("tests").child(splitFile(suite.file).name)
    var sources: seq[string]
    for source in c.child("sources").strings & suiteOptions.child("sources").strings:
      sources.add(rooted(projectRoot, source))
    if c.child("sources") != nil and c.child("sources").kind == JArray:
      for source in c.child("sources"):
        if source.kind == JObject and source.hasKey("path"):
          sources.add(rooted(projectRoot, source["path"].getStr()))
    var includePaths: seq[string]
    for path in c.child("include").strings: includePaths.add(rooted(projectRoot, path))
    let source = readFile(suite.file)
    let diagnostics = newEngine()
    diagnostics.setSource(source, suite.file)
    var program = newParser(newLexer(source, diagnostics).lex(), diagnostics).parse()
    if diagnostics.failed:
      raise newException(ValueError, renderAll(diagnostics.messages(),
        RenderOptions(style: "short", color: false)))
    if not synthesizeTests(program, suite): raise newException(ValueError, "Test suite was not found: " & suite.name)
    expand(program, diagnostics, projectRoot,
      cBinding.BindOptions(includePaths: includePaths))
    let resolution = newResolver(diagnostics, projectRoot).resolve(program, suite.file)
    program = link(program, resolution)
    lint(program, diagnostics, backend)
    let level = manifest.child("requires").text("base")
    capabilities(program, diagnostics, level)
    let checker = newChecker(diagnostics)
    checker.check(program)
    if diagnostics.failed:
      raise newException(ValueError, renderAll(diagnostics.messages(),
        RenderOptions(style: "short", color: false)))
    let module = monomorphize(lower(program, checker.types))
    let outDir = projectRoot / ".artifacts" / "test" /
      (splitFile(suite.file).name & "_" & suite.name).replace(' ', '_')
    createDir(outDir)
    let cCompiler = if getEnv("CC").len > 0: getEnv("CC") elif findExe("clang").len > 0: findExe("clang") else: "cc"
    ensure(level)
    let nativeConfig = build.child("native")
    let native = buildOptions.Native(name: "test", kind: "exe",
      compile: true, level: level,
      compiler: if backend == "c": cCompiler else: "",
      runtime: build.child("runtime").text(),
      substrate: build.child("substrate").text(),
      native: buildOptions.NativeBinding(
        substrate: nativeConfig.child("substrate").text(),
        clobbers: nativeConfig.child("clobbers").strings),
      cpu: build.child("cpu").text(), coverage: build.child("coverage").text(),
      sanitize: build.child("sanitize").text(), docs: build.child("docs").flag,
      sources: sources, includePaths: includePaths, flags: c.child("flags").strings,
      libs: linkOptions.child("libs").strings,
      frameworks: linkOptions.child("frameworks").strings,
      cpp: linkOptions.child("cpp").flag)
    var success = false
    var artifactPath, buildError: string
    if backend == "c":
      let built = cDriver.build(module, "dev", outDir, native, suite.file)
      success = built.success; artifactPath = built.artifact; buildError = built.error
    else:
      let zig = detect().path
      if zig.len == 0: raise newException(IOError, "Zig backend is missing; install the pinned toolchain")
      let built = zigDriver.build(module, "dev", outDir, zig, suite.file, native)
      success = built.success; artifactPath = built.artifact; buildError = built.error
    if not success: raise newException(OSError, if buildError.len > 0: buildError else: "Test build failed")
    let execution = execCmdEx(quoteShell(artifactPath), workingDir = projectRoot)
    result.passed = execution.exitCode == 0
    result.output = execution.output
    result.files = @[suite.file] & sources & (if fileExists(manifestPath): @[manifestPath] else: @[])
    result.opaque = sources.len > 0 or native.libs.len > 0
    if not result.passed: result.error = "Test process exited with " & $execution.exitCode
  except CatchableError as error:
    result.passed = false
    result.error = error.msg
  result.timeMs = int((epochTime() - started) * 1000)

proc runTests*(root = getCurrentDir(); filter = ""; backend = "zig"; executor: TestExecutor = nil): seq[TestResult] =
  var suites = discoverTests(root)
  if filter.len > 0: suites = suites.filterIt(filter in it.name)
  for suite in suites: result.add(runSingleTest(suite, backend, executor))

proc watchTests*(root = getCurrentDir(); filter = ""; backend = "zig"; executor: TestExecutor = nil): TestWatcher =
  TestWatcher(root: absolutePath(root), filter: filter, backend: backend, executor: executor,
    snapshot: initTable[string, string]())

proc poll*(watcher: TestWatcher): seq[TestResult] =
  if watcher == nil or watcher.stopped: return @[]
  var changed = watcher.snapshot.len == 0
  var current = initTable[string, string]()
  for suite in discoverTests(watcher.root):
    let stamp = $getLastModificationTime(suite.file)
    current[suite.file] = stamp
    if watcher.snapshot.getOrDefault(suite.file) != stamp: changed = true
  watcher.snapshot = current
  if changed: result = runTests(watcher.root, watcher.filter, watcher.backend, watcher.executor)

proc stop*(watcher: TestWatcher) =
  if watcher != nil: watcher.stopped = true
